marimo/_save/cache.py

```diff
@@ -6,36 +6,102 @@
 from dataclasses import dataclass
 from typing import TYPE_CHECKING, Any, Literal
 
+from marimo._plugins.ui._core.ui_element import UIElement
 from marimo._runtime.context import get_context
+from marimo._runtime.state import SetFunctor
 
 if TYPE_CHECKING:
+    from types import FrameType
+
     from marimo._ast.visitor import Name
 
-CacheType = Literal["ContentAddressed", "ExecutionPath", "Unknown"]
+CacheType = Literal[
+    "ContextExecutionPath",
+    "ContentAddressed",
+    "ExecutionPath",
+    "Pure",
+    "Unknown",
+]
 # Easy visual identification of cache type.
 CACHE_PREFIX: dict[CacheType, str] = {
+    "ContextExecutionPath": "X_",
     "ContentAddressed": "C_",
     "ExecutionPath": "E_",
+    "Pure": "P_",
     "Unknown": "U_",
 }
 
 ValidCacheSha = namedtuple("ValidCacheSha", ("sha", "cache_type"))
 
 
+# BaseException because "raise _ as e" is utilized.
+class CacheException(BaseException):
+    pass
+
+
 @dataclass
 class Cache:
     defs: dict[Name, Any]
     hash: str
+    stateful_refs: set[str]
     cache_type: CacheType
     hit: bool
 
+    def restore(self, frame: FrameType) -> None:
+        for lookup, var in self.contextual_defs():
+            frame.f_locals[var] = self.defs[lookup]
+
+        defs = {**globals(), **frame.f_locals}
+        for ref in self.stateful_refs:
+            if ref not in defs:
+                raise CacheException(
+                    "Failure while restoring cached values. "
+                    "Cache expected a reference to a "
+                    f"variable that is not present ({ref})."
+                )
```

Comment:
Does a Cached exception leave the frame in a weird state, since variables have already been restored? Do they need to be un-restored, or is it fine? My guess is fine but just double checking ...

Reply:
It will be the same as if the cell failed. The `with` statement cannot be nested under a try catch- so it will kick back out to runner which will clean up the defs